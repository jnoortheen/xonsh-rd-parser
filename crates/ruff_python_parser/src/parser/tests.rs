use crate::{Mode, ParseOptions, parse, parse_expression, parse_module};

#[test]
fn test_modes() {
    let source = "a[0][1][2][3][4]";

    assert!(parse(source, ParseOptions::from(Mode::Expression)).is_ok());
    assert!(parse(source, ParseOptions::from(Mode::Module)).is_ok());
}

#[test]
fn test_expr_mode_invalid_syntax1() {
    let source = "first second";
    let error = parse_expression(source).unwrap_err();

    insta::assert_debug_snapshot!(error);
}

#[test]
fn test_expr_mode_invalid_syntax2() {
    let source = r"first

second
";
    let error = parse_expression(source).unwrap_err();

    insta::assert_debug_snapshot!(error);
}

#[test]
fn test_expr_mode_invalid_syntax3() {
    let source = r"first

second

third
";
    let error = parse_expression(source).unwrap_err();

    insta::assert_debug_snapshot!(error);
}

#[test]
fn test_expr_mode_valid_syntax() {
    let source = "first

";
    let parsed = parse_expression(source).unwrap();

    insta::assert_debug_snapshot!(parsed.expr());
}

#[test]
fn test_unicode_aliases() {
    // https://github.com/RustPython/RustPython/issues/4566
    let source = r#"x = "\N{BACKSPACE}another cool trick""#;
    let suite = parse_module(source).unwrap().into_suite();

    insta::assert_debug_snapshot!(suite);
}

#[test]
fn test_xonsh_procs() {
    let source = r#"$(echo 123 'hello')"#;
    let suite = parse_module(source).unwrap().into_suite();

    insta::assert_debug_snapshot!(suite);
}

#[test]
fn test_fstring_expr_inner_line_continuation_and_t_string() {
    let source = r#"f'{\t"i}'"#;

    let parsed = parse_expression(source);

    let error = parsed.unwrap_err();

    insta::assert_debug_snapshot!(error);
}

#[test]
fn test_fstring_expr_inner_line_continuation_newline_t_string() {
    let source = r#"f'{\
t"i}'"#;

    let parsed = parse_expression(source);

    let error = parsed.unwrap_err();

    insta::assert_debug_snapshot!(error);
}
