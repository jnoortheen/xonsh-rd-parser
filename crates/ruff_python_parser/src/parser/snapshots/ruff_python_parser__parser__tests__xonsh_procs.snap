---
source: crates/ruff_python_parser/src/parser/tests.rs
expression: suite
snapshot_kind: text
---
[
    Expr(
        StmtExpr {
            range: 0..19,
            value: Call(
                ExprCall {
                    range: 0..19,
                    func: Attribute(
                        ExprAttribute {
                            range: 0..19,
                            value: Call(
                                ExprCall {
                                    range: 0..19,
                                    func: Attribute(
                                        ExprAttribute {
                                            range: 2..6,
                                            value: Name(
                                                ExprName {
                                                    range: 2..6,
                                                    id: Name("__xonsh__"),
                                                    ctx: Load,
                                                },
                                            ),
                                            attr: Identifier {
                                                id: Name("cmd"),
                                                range: 2..6,
                                            },
                                            ctx: Load,
                                        },
                                    ),
                                    arguments: Arguments {
                                        range: 2..19,
                                        args: [
                                            StringLiteral(
                                                ExprStringLiteral {
                                                    range: 2..6,
                                                    value: StringLiteralValue {
                                                        inner: Single(
                                                            StringLiteral {
                                                                range: 2..6,
                                                                value: "echo",
                                                                flags: StringLiteralFlags {
                                                                    quote_style: Single,
                                                                    prefix: Empty,
                                                                    triple_quoted: false,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                            StringLiteral(
                                                ExprStringLiteral {
                                                    range: 7..10,
                                                    value: StringLiteralValue {
                                                        inner: Single(
                                                            StringLiteral {
                                                                range: 7..10,
                                                                value: "123",
                                                                flags: StringLiteralFlags {
                                                                    quote_style: Single,
                                                                    prefix: Empty,
                                                                    triple_quoted: false,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                            StringLiteral(
                                                ExprStringLiteral {
                                                    range: 11..18,
                                                    value: StringLiteralValue {
                                                        inner: Single(
                                                            StringLiteral {
                                                                range: 11..18,
                                                                value: "hello",
                                                                flags: StringLiteralFlags {
                                                                    quote_style: Single,
                                                                    prefix: Empty,
                                                                    triple_quoted: false,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                        ],
                                        keywords: [],
                                    },
                                },
                            ),
                            attr: Identifier {
                                id: Name("out"),
                                range: 0..19,
                            },
                            ctx: Load,
                        },
                    ),
                    arguments: Arguments {
                        range: 0..19,
                        args: [],
                        keywords: [],
                    },
                },
            ),
        },
    ),
]
